hazelcast/src/main/java/com/hazelcast/quorum/impl/ProbabilisticQuorumFunction.java

```diff
@@ -0,0 +1,165 @@
+ * limitations under the License.
+ */
+
+package com.hazelcast.quorum.impl;
+
+import com.hazelcast.core.Member;
+import com.hazelcast.core.MemberAttributeEvent;
+import com.hazelcast.core.MembershipEvent;
+import com.hazelcast.core.MembershipListener;
+import com.hazelcast.internal.cluster.fd.PhiAccrualClusterFailureDetector;
+import com.hazelcast.quorum.HeartbeatAware;
+import com.hazelcast.quorum.QuorumFunction;
+import com.hazelcast.util.Clock;
+
+import java.util.Collection;
+
+public class ProbabilisticQuorumFunction extends AbstractPingAwareQuorumFunction
+        implements HeartbeatAware, QuorumFunction, MembershipListener {
+
+    private final double suspicionThreshold;
+    private final int quorumSize;
+    private final int maxSampleSize;
+    private final long minStdDeviationMillis;
+    private final long acceptableHeartbeatPauseMillis;
+    private final long heartbeatIntervalMillis;
+    private final PhiAccrualClusterFailureDetector failureDetector;
+
+    public ProbabilisticQuorumFunction(int quorumSize, long heartbeatIntervalMillis,
+                                       long acceptableHeartbeatPauseMillis,
+                                       int maxSampleSize,
+                                       long minStdDeviationMillis,
+                                       double suspicionThreshold) {
+        this.heartbeatIntervalMillis = heartbeatIntervalMillis;
+        this.acceptableHeartbeatPauseMillis = acceptableHeartbeatPauseMillis;
+        this.maxSampleSize = maxSampleSize;
+        this.minStdDeviationMillis = minStdDeviationMillis;
+        this.suspicionThreshold = suspicionThreshold;
+        this.quorumSize = quorumSize;
+
+        this.failureDetector = new PhiAccrualClusterFailureDetector(acceptableHeartbeatPauseMillis,
+                heartbeatIntervalMillis, suspicionThreshold, maxSampleSize, minStdDeviationMillis);
+    }
+
+    @Override
+    public boolean apply(Collection<Member> members) {
+        if (members.size() < quorumSize) {
+            return false;
+        }
+
+        int count = 0;
+        long timestamp = Clock.currentTimeMillis();
+        for (Member member : members) {
+            if (!isAlivePerIcmp(member)) {
+                continue;
+            }
+            if (failureDetector.isAlive(member, timestamp)) {
+                count++;
+            }
+        }
+        return count >= quorumSize;
+    }
+
+    @Override
+    public void memberAdded(MembershipEvent membershipEvent) {
+        // ensure ping FD has heard at least once from each member
+        if (pingFDEnabled) {
+            pingFailureDetector.heartbeat(membershipEvent.getMember());
+        }
+    }
+
+    @Override
+    public void memberRemoved(MembershipEvent membershipEvent) {
+        if (pingFDEnabled) {
+            pingFailureDetector.remove(membershipEvent.getMember());
+            failureDetector.remove(membershipEvent.getMember());
```

Comment:
Shouldn't `failureDetector.remove(membershipEvent.getMember())` line be outside of `if` block?